l3backend/l3backend-color.dtx

```diff
@@ -1248,6 +1253,401 @@
+      color = ""
+    end
+  end
+%    \end{macrocode}
+%   Loop done, set up the list.
+%    \begin{macrocode}
+  if get_type(list) == node_map.list then
+    list.head = head
+  else
+    list.replace = head
+  end
+  return color
+end
+%    \end{macrocode}
+% \end{macro}
+%
+% Activate the code: we assume that the |pre_shipout_filter| callback is
+% available as this is pretty burning-edge stuff.
+%    \begin{macrocode}
+if luatexbase.callbacktypes.pre_shipout_filter then
+  luatexbase.add_to_callback(
+    "pre_shipout_filter",
+    function(list)
+      traverse(list,"")
+      return true
+    end,
+    "color")
+end
+%    \end{macrocode}
+%
+% For recent versions of \pkg{luaotfload}, we can register a callback to
+% control how coloring glyph is handled for the color feature.
+%    \begin{macrocode}
+if luaotfload and luaotfload.set_colorhandler then
+  local set_attribute = node.direct.set_attribute
+  luaotfload.set_colorhandler(function(head,n,color)
+    set_attribute(n,attribute,color_index(color))
+    return head,n
+  end)
+end
+%    \end{macrocode}
+%
+% \begin{macro}{luacmd}
+%   Copied from \pkg{l3luatex}: we need to think about how to make this available
+%   (perhaps in |ltx.utils|?).
+%    \begin{macrocode}
+local luacmd do
+  local token_create = token.create
+  local set_lua = token.set_lua
+  local undefined_cs = token.command_id'undefined_cs'
+
+  if not context and not luatexbase then require'ltluatex' end
+  if luatexbase then
+    local new_luafunction = luatexbase.new_luafunction
+    local functions = lua.get_functions_table()
+    function luacmd(name, func, ...)
+      local id
+      local tok = token_create(name)
+      if tok.command == undefined_cs then
+        id = new_luafunction(name)
+        set_lua(name, id, ...)
+      else
+        id = tok.index or tok.mode
+      end
+      functions[id] = func
+    end
+  elseif context then
+    local register = context.functions.register
+    local functions = context.functions.known
+    function luacmd(name, func, ...)
+      local tok = token.create(name)
+      if tok.command == undefined_cs then
+        token.set_lua(name, register(func), ...)
+      else
+        functions[tok.index or tok.mode] = func
+      end
+    end
+  end
+end
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}{\tex_pdfxform:D}
+%   We redefine the \pkg{expl3} version of the primitive to do the replacement
+%   implicitly. The keywords \texttt{margin} and \texttt{type} are not
+%   documented at the \TeX{} level.
+%    \begin{macrocode}
+luacmd('tex_pdfxform:D', function()
+  local attr = scan_keyword'attr' and scan_string() or nil
+  local resources = scan_keyword'length' and scan_string() or nil
+  local xtype = scan_keyword'type' and scan_string() or nil
+  local margin = scan_keyword'margin' and scan_string() or nil
+  local n = scan_int() or nil
+  traverse(getbox(n),"")
+  saveboxresources(n,attr,resources,false,xtype,margin)
+end,'global','protected')
```

Comment:
The XForm issue...

This breaks *every* use of `\pdf_xform:D` in TeX Live. :laughing:  (Since it doesn't work with `\tex_immediate:D`) Properly supporting `\tex_immediate:D` here is complicated, as soon as we are non-expandable immediate breaks, if we are expandable it doesn't behave like the primitive and supporting immediate is still a pain. The mess in luacolor is the closest solution I can think of, but even that is far from perfect and probably inappropriate for kernel code.

So I think the only reasonable way forward is to change the name and provide a second name for the immediate version. If we use non-expl3 names (or aliases) and define them with also for pdfTeX is should be relatively simple to get most users to change. Since many uses of this happen in generic code, we have to make sure not to depend on LaTeX there, so I would go for just the wrapped primitive.

Anyway, this is a much more general issue than just for colors.

Reply:
I wasn't sure how `\immediate` interacted with the Lua here - sounds like 'not properly'.

In any case, we don't necessarily want/need to worry about that. I very deliberately only set `\tex_pdfxfom:D`, meaning that the only place I'm expecting this token to be used is from within `l3backend`. At the moment, `l3color` is pretty specialised and I think I'd be happy with a position of

> If you use `l3color` inside an XForm, it's all got to be handled by `expl3`.

So we could then define some `\__kernel_box_xform:w` function(s) and use them  - no messing with other people's code.

Reply:
The problem I see with

> If you use l3color inside an XForm, it's all got to be handled by expl3.

is that you often can't really influence if your code is used in an XForm. E.g. some user using code in a TikZ label shouldn't have to worry about if the code uses l3color or not and if the TikZ document happens to use `expl3` and needs XForms (Or even what XForms are).

Another possible solution to the original issue: We could avoid the problem if we are open to redefining `\immediate` itself. A custom `\immediate` emulation would make `\immediate` compatible with Lua commands, but it's meaning would become `luacall n` and tracing would show an additional `luacall` everytime `\immediate` is used.

Reply:
> Another possible solution to the original issue: We could avoid the problem if we are open to redefining `\immediate` itself. A custom `\immediate` emulation would make `\immediate` compatible with Lua commands, but it's meaning would become `luacall n` and tracing would show an additional `luacall` everytime `\immediate` is used.

I wouldn't be surprised at all if that is hitting us in the face. Maybe a `\protected` version would do, but it sounds dangerous.

Reply:
> > Another possible solution to the original issue: We could avoid the problem if we are open to redefining `\immediate` itself. A custom `\immediate` emulation would make `\immediate` compatible with Lua commands, but it's meaning would become `luacall n` and tracing would show an additional `luacall` everytime `\immediate` is used.
> 
> I wouldn't be surprised at all if that is hitting us in the face. Maybe a `\protected` version would do, but it sounds dangerous.

In which way? Do you think that people actually test for `\immediate` being the primitive or do you think that it would break unrelated code since it would behave differently"

Reply:
> The problem I see with
> 
> > If you use l3color inside an XForm, it's all got to be handled by expl3.
> 
> is that you often can't really influence if your code is used in an XForm. E.g. some user using code in a TikZ label shouldn't have to worry about if the code uses l3color or not and if the TikZ document happens to use `expl3` and needs XForms (Or even what XForms are).

At least at present, I'm seeing `l3color` as a definite 'opt in when you know you want it', for things like `l3draw`. As such, I don't expect anything using `l3color` to be used inside an arbitrary colour-dependent context: anything inside `pgf` needs to be using classical `color` I'd say. Even allowing `pgf`, there are not a lot of use cases for XForms that use text: the obvious one is for graphics inclusion, and that's not subject to Lua-based colour.

Of course, we can decide to provide some more abstract interfaces for XForms, but they'll need to be in another place:  if we want arbitrary code to use them, they've got to work in plain (as `pgf` does), and that means not in `expl3` (as plain users might load `graphics.tex` but will likely not go for `expl3` in a routine sense).

Reply:
Right. 

You're right.

> So we could then define some `\__kernel_box_xform:w` function(s) and use them - no messing with other people's code.

Namewise I think it should include `pdf` since it's a rather PDF specific concept.